nccl flight recorder

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110960

Keep a buffer of the last 16384 nccl work actions, including the stack
trace that launched the event.

When torch._C._distributed_c10d._dump_nccl_trace(), it an dump these to
a pickled archive.

For each action we get:
process_group_id, seq_id, collective_name, size_of_first_tensor, stack trace

state - issued, started, completed (based on cuda events and queried if
necessary when the dump is requested)

I tested that it is possible to query event state when the streams are
otherwise stuck.

Differential Revision: [D50138956](https://our.internmc.facebook.com/intern/diff/D50138956)

zdevito committed:
nccl flight recorder

Keep a buffer of the last 16384 nccl work actions, including the stack
trace that launched the event.

When torch._C._distributed_c10d._dump_nccl_trace(), it an dump these to
a pickled archive.

For each action we get:
process_group_id, seq_id, collective_name, size_of_first_tensor, stack trace

state - issued, started, completed (based on cuda events and queried if
necessary when the dump is requested)

I tested that it is possible to query event state when the streams are
otherwise stuck.

[ghstack-poisoned]

zdevito committed:
Update on "nccl flight recorder"


Keep a buffer of the last 16384 nccl work actions, including the stack
trace that launched the event.

When torch._C._distributed_c10d._dump_nccl_trace(), it an dump these to
a pickled archive.

For each action we get:
process_group_id, seq_id, collective_name, size_of_first_tensor, stack trace

state - issued, started, completed (based on cuda events and queried if
necessary when the dump is requested)

I tested that it is possible to query event state when the streams are
otherwise stuck.

[ghstack-poisoned]

zdevito committed:
Update on "nccl flight recorder"


Keep a buffer of the last 16384 nccl work actions, including the stack
trace that launched the event.

When torch._C._distributed_c10d._dump_nccl_trace(), it an dump these to
a pickled archive.

For each action we get:
process_group_id, seq_id, collective_name, size_of_first_tensor, stack trace

state - issued, started, completed (based on cuda events and queried if
necessary when the dump is requested)

I tested that it is possible to query event state when the streams are
otherwise stuck.

[ghstack-poisoned]

zdevito commented:
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50138956).

malfet commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -283,6 +289,198 @@ inline void errorIfCapturingNonCapturableNCCL(c10::cuda::CaptureStatus status) {
+    pickler.pushIValue(v);
+    pickler.stop();
+  }
+  return std::string(result.begin(), result.end());
```
Cool, I didn't know that C++ strings does not have to be null terminated:
https://godbolt.org/z/M7ve7dTxb

malfet commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -283,6 +289,198 @@ inline void errorIfCapturingNonCapturableNCCL(c10::cuda::CaptureStatus status) {
+  return std::string(result.begin(), result.end());
+}
+c10::Dict<c10::IValue, c10::IValue> new_dict() {
+  return c10::Dict<c10::IValue, c10::IValue>(
+      c10::AnyType::get(), c10::AnyType::get());
```
Nit
C++ should auto-create an object, shouldn't it, i.e. 
```cpp
std::unordered_map<int, int> foo() {
    return {{1, 2}};
}
```
```suggestion
  return {{c10::AnyType::get(), c10::AnyType::get()}};
```

zdevito replied:
ctor's are marked explicit, so it doesn't work.

malfet commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -283,6 +289,198 @@ inline void errorIfCapturingNonCapturableNCCL(c10::cuda::CaptureStatus status) {
+      c10::AnyType::get(), c10::AnyType::get());
+}
+c10::List<c10::IValue> new_list() {
+  return c10::List<c10::IValue>(c10::AnyType::get());
```
Nit (same as above)
```suggestion
  return {{c10::AnyType::get()};
```

malfet commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -283,6 +289,198 @@ inline void errorIfCapturingNonCapturableNCCL(c10::cuda::CaptureStatus status) {
+    c10::SmallVector<int64_t, 4> sizes_;
+    std::shared_ptr<torch::CapturedTraceback> traceback_;
+    EventList *start_, *end_;
+    const char* state_ = "scheduled";
```
Looks like this should be an enum, shouldn't it?

zdevito replied:
It never gets used as such yet and just gets turned back directly into the string, so I didn't bother yet.

malfet commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -283,6 +289,198 @@ inline void errorIfCapturingNonCapturableNCCL(c10::cuda::CaptureStatus status) {
+    auto traceback = torch::CapturedTraceback::gather(true, true, true);
+    std::lock_guard<std::mutex> guard(mutex_);
+    at::IntArrayRef sizes;
+    if (inputs && inputs->size() > 0) {
```
```suggestion
    static const std::vector<at::Tensor> emptyVector;
    if (!inputs.value_or(emptyVector).empty()) {
```

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -283,6 +289,198 @@ inline void errorIfCapturingNonCapturableNCCL(c10::cuda::CaptureStatus status) {
+      dict.insert(sizes_s, at::IntArrayRef(e.sizes_.begin(), e.sizes_.end()));
+      dict.insert(state_s, e.state_);
+      auto frames = new_list();
+      for (int64_t frame : tb) {
```
i'm confused about the different frame/traceback data structures

is frame an index (would it match `i` above)?  or is it some kind of 'frame id'?

trying to see if we can simplify this..

zdevito replied:
The complexity here makes sure we only create 1 dictionary per unique frame entry across all tracebacks, which will keep the eventual size of the pickled data  small. `all_frames` is a flat list of every frame we encountered and then each traceback is a list of offsets into all_frames.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -283,6 +289,198 @@ inline void errorIfCapturingNonCapturableNCCL(c10::cuda::CaptureStatus status) {
+    }
+  }
+
+  std::string dump() {
```
bikeshedding: dump and report both kind of sound like synonyms to me.  report actually doesn't 'generate a report' it just updates the state of the state of the events inside the buffer to the latest info it can get.  Maybe leave dump as is but change report to something more descriptive?

zdevito replied:
I named report `dump_entries`. It really only exists to have a function that lives inside of the lock, and another that lives outside of it. `symbolize` is expensive, so it needs to be outside that lock.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -283,6 +289,198 @@ inline void errorIfCapturingNonCapturableNCCL(c10::cuda::CaptureStatus status) {
 } // namespace
 
+namespace {
+std::string write_pickle(const c10::IValue& v) {
```
nit: maybe name 'pickle_str' or something? people might expect this to create a file

wconstab reviewed (commented):
looks good to me! 

have you tried it on anything yet to see how it works in practice?  any thoughts on adding a test for it?

zdevito committed:
Update on "nccl flight recorder"


Keep a buffer of the last 16384 nccl work actions, including the stack
trace that launched the event.

When torch._C._distributed_c10d._dump_nccl_trace(), it an dump these to
a pickled archive.

For each action we get:
process_group_id, seq_id, collective_name, size_of_first_tensor, stack trace

state - issued, started, completed (based on cuda events and queried if
necessary when the dump is requested)

I tested that it is possible to query event state when the streams are
otherwise stuck.

Differential Revision: [D50138956](https://our.internmc.facebook.com/intern/diff/D50138956)

[ghstack-poisoned]

zdevito commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -283,6 +289,198 @@ inline void errorIfCapturingNonCapturableNCCL(c10::cuda::CaptureStatus status) {
+    for (auto& e : result) {
+      tracebacks.push_back(e.traceback_.get());
+    }
+    torch::SymbolizedTracebacks stracebacks = torch::symbolize(tracebacks);
```
tracebacks are lists of void* and PyCode objects, this turns them into strings. It is how we keep gathering the tracebacks so cheap.

zdevito commented:
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50138956).

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -327,48 +325,69 @@ struct NCCLTraceBuffer {
-    c10::SmallVector<int64_t, 4> sizes_;
+
     std::shared_ptr<torch::CapturedTraceback> traceback_;
+    // we borrow pointser to start_ and end_ so we can query the state
```
typo

zdevito committed:
Update on "nccl flight recorder"


Keep a buffer of the last 16384 nccl work actions, including the stack
trace that launched the event.

When torch._C._distributed_c10d._dump_nccl_trace(), it an dump these to
a pickled archive.

For each action we get:
process_group_id, seq_id, collective_name, size_of_first_tensor, stack trace

state - issued, started, completed (based on cuda events and queried if
necessary when the dump is requested)

I tested that it is possible to query event state when the streams are
otherwise stuck.

Differential Revision: [D50138956](https://our.internmc.facebook.com/intern/diff/D50138956)

[ghstack-poisoned]

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3347,6 +3349,206 @@ def test_ddp_set_sparse_metadata(self):
+        device = self.local_device
+        a = torch.full((3, 4), float(self.rank), device=device)
+        for i in range(2):
+            # test some other primitives to make sure
```
i expected to see some assertion about the op type in the checks below, did you forget to add that part?

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3347,6 +3349,206 @@ def test_ddp_set_sparse_metadata(self):
+    @skip_but_pass_in_sandcastle_if(torch.cuda.device_count() < 2, "NCCL test requires 2+ GPUs")
+    def test_trace_while_active(self):
+        if self.rank == self.MAIN_PROCESS_RANK:
+            for c in self.children_pipes:
```
is this just a barrier to make sure main process doesn't exit before children exit?

hmm. wonder if we should hook this up as a standard part of the testing infra, it seems like most people would want this but maybe didn't think about how to do it.

wconstab replied:
oh nevermind. its a way to use the parent as a barrier between rank0 and ranks!==0

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3347,6 +3349,206 @@ def test_ddp_set_sparse_metadata(self):
+                th.start()
+                # fill the cuda buffer, at around 1024 events
+                # this will stall
+                for i in range(2000):
```
what's the significance of filling up the cuda buffer?  iiuc the second allreduce called on rank != 0 is sufficient to cause a stall.

wconstab commented:
changes lgtm. going to run an internal test before stamping.

zdevito commented:
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50138956).

zdevito committed:
Update on "nccl flight recorder"


Keep a buffer of the last 16384 nccl work actions, including the stack
trace that launched the event.

When torch._C._distributed_c10d._dump_nccl_trace(), it an dump these to
a pickled archive.

For each action we get:
process_group_id, seq_id, collective_name, size_of_first_tensor, stack trace

state - issued, started, completed (based on cuda events and queried if
necessary when the dump is requested)

I tested that it is possible to query event state when the streams are
otherwise stuck.

Differential Revision: [D50138956](https://our.internmc.facebook.com/intern/diff/D50138956)

[ghstack-poisoned]

zdevito commented:
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50138956).

zdevito commented:
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50138956).
